core/domain/change_domain.py

```diff
@@ -253,6 +262,9 @@ def __getattr__(self, name: str) -> str:
         # AttributeError needs to be thrown in order to make
         # instances of this class picklable.
         try:
-            return self.__dict__[name]
+            # Internally __dict__() method is annotated with Dict[str, Any]
+            # return type. So to narrow down the return type from Any to str
+            # we used cast here.
+            return cast(str, self.__dict__[name])
```

Comment:
Done!

Reply:
Due to the addition of assertion here, Backend coverage fails. And I'm a bit confused about where to add test for this because there is no test file for `change_domain.py`

Reply:
link for the failing test: (https://github.com/oppia/oppia/runs/6984319705?check_suite_focus=true#step:13:83). 
I also tried to fix backend coverage with the explicit assignment 
```
            dict_name: str = self.__dict__[name]
            return dict_name
```
but it also not working.
@vojtechjelinek PTAL!

Reply:
Maybe you will need to add a new test file.

Reply:
While writing tests I found that `__getattr__` is only defined for throwing the `AttributeError`. Because:
In method `to_dict()`, we are calling `getattr()` only  when we are sure that the attribute exists on `self` by using `hasattr()`.
```
            if hasattr(self, attribute_name):
                base_change_dict[attribute_name] = getattr(
                    self, attribute_name)
```
But if for some reason `getattr` is not able to fetch that attribute, it calls `__getattr__` before throwing an error and here we are doing `self.__dict__[name]` to raise and catch the exception. So, Lines written after `self.__dict__[name]` is never executed. So, that's why `assert isinstance()` is never covered in coverage.
```
    def __getattr__(self, name: str) -> str:
        # AttributeError needs to be thrown in order to make
        # instances of this class picklable.
        try:
            dict_name = self.__dict__[name] ------------- executed and  exception is raised
            assert isinstance(dict_name, str) ------------- this is never executed because exception is raised before and control is transferred to except block.
            return dict_name -------- never executed.
        except KeyError as e:
            raise AttributeError(name) from e
```
While cast() is covered in coverage because it comes under the same line where the exception is raised.
```
    def __getattr__(self, name: str) -> str:
        # AttributeError needs to be thrown in order to make
        # instances of this class picklable.
        try:
            return cast(str, self.__dict__[name]) ------------- executed and  exception is raised
        except KeyError as e:
            raise AttributeError(name) from e
```
Also, we cannot remove this `__getattr__` because this helps MyPy while reading for attributes. Because in `__init__` method we are setting attributes dynamically.

Reply:
@vojtechjelinek PTAL!

Reply:
Okay, please explain this in the comment,

Reply:
Done!